usaspending_api/awards/migrations/0075_financialaccountsbyawards_distinct_award_key.py

```diff
@@ -0,0 +1,18 @@
+# Generated by Django 2.2.13 on 2020-08-31 15:03
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('awards', '0074_financialaccountsbyawards_distinct_award_key'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='financialaccountsbyawards',
+            name='distinct_award_key',
+            field=models.TextField(db_index=True),
+        ),
+    ]
```

Comment:
Will this set some kind of default value before the update is run? In my experience running these kinds of migrations locally prompts the user for a default value, and I'm unsure how that would work when running this migration via Jenkins.

Reply:
There are two migrations, the first allows nulls (so the default value is null) and this migration adds the not null constraint. It is expected that a backfill script would run between the migrations. This is a common pattern for adding not null columns to production databases

Reply:
> Since all 4 contributing fields can be null

Are we sure that is the case? Currently there are no records missing all 4 fields. I haven't checked the DAIMS to confirm.

AFAIK the a fresh DB will not contain any records in `submission_attributes`, so what would the migration do in this case? When pre-existing data exists, there is a backfill script in the PR to run.

I strongly recommend never having data be populated by a migration unless is it **very** static data, like status codes for a process. The concern for a developer's local DB is a one-time event. But these migrations are run hundreds to thousands of times by CI tests each week. We're already having issues with the DABS submission window data inserted by a migration. It adds time and messes with the idea that the test environment db is a "clean slate"

Reply:
In my mind, migrations should be migrations of a production system ... not a developer first-time setup. So they should consider how to migrate a db from x -> y if there is production data in state "x". This case wouldn't be adding any data to the system, it is a backfill only operating on existing data. On a clean unit test database, the migration would be a no-op since there's no `financial_accounts_by_awards` data at migration-time in that scenario.

As you mentioned this is a common scenario: "how to add a `NOT NULL` column to an existing table with data". I usually see the migrations for this orchestrating all 3 steps though, ideally in a transaction if possible in case any one step fails.
1. add the column as NULL allowed
2. add the missing data
3. make the column NOT NULL

Still, if this is the pattern we have, where migrations are strictly about schema and not about data, **then we should just stick with it**. Benefit: consistency. Downside: can't just run all migrations on an existing and have to consult one or more deployment coordination guides.